go-controller/pkg/ovn/host_subnet_allocator.go

```diff
@@ -0,0 +1,181 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/subnetallocator"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+type hostSubnetAllocator struct {
+	// Don't inherit from BaseSubnetAllocator to ensure users of
+	// hostSubnetAllocator can't directly call the underlying methods
+	base *subnetallocator.BaseSubnetAllocator
+}
+
+func NewHostSubnetAllocator() *hostSubnetAllocator {
+	return &hostSubnetAllocator{
+		base: subnetallocator.NewSubnetAllocator(),
+	}
+}
+
+func (sna *hostSubnetAllocator) InitRanges(subnets []config.CIDRNetworkEntry) error {
+	for _, entry := range subnets {
+		if err := sna.base.AddNetworkRange(entry.CIDR, entry.HostSubnetLength); err != nil {
+			return err
+		}
+		klog.V(5).Infof("Added network range %s to host subnet allocator", entry.CIDR)
+	}
+
+	// update metrics for host subnets
+	v4count, _, v6count, _ := sna.base.Usage()
+	metrics.RecordSubnetCount(float64(v4count), float64(v6count))
+	return nil
+}
+
+// MarkSubnetsAllocated will mark the given subnets as already allocated by
+// the given owner. Marking is all-or-nothing; if marking one of the subnets
+// fails then none of them are marked as allocated.
+func (sna *hostSubnetAllocator) MarkSubnetsAllocated(node *kapi.Node, subnets []*net.IPNet) error {
+	if err := sna.base.MarkAllocatedNetworks(subnets, node.Name); err != nil {
+		return err
+	}
+	_, v4used, _, v6used := sna.base.Usage()
+	metrics.RecordSubnetUsage(float64(v4used), float64(v6used))
+	return nil
+}
+
+func (sna *hostSubnetAllocator) AllocateNodeSubnets(node *kapi.Node, ipv4Mode, ipv6Mode bool) ([]*net.IPNet, []*net.IPNet, error) {
+	hostSubnets, err := util.ParseNodeHostSubnetAnnotation(node)
+	if err != nil {
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s host subnets annotations: %v", node.Name, err)
+	}
+	allocatedSubnets := []*net.IPNet{}
+
+	// OVN can work in single-stack or dual-stack only.
+	currentHostSubnets := len(hostSubnets)
+	expectedHostSubnets := 1
+	// if dual-stack mode we expect one subnet per each IP family
+	if ipv4Mode && ipv6Mode {
+		expectedHostSubnets = 2
+	}
+
+	// node already has the expected subnets annotated
+	// assume IP families match, i.e. no IPv6 config and node annotation IPv4
+	if expectedHostSubnets == currentHostSubnets {
```

Comment:
shouldn't we check for chances of hostSubnets duplicates with other nodes here ?

Reply:
@pperiyasamy yep, you were right. Addressed in the last commit now.